apollo-router/src/plugins/telemetry/fmt_layer.rs

```diff
@@ -167,28 +167,56 @@ where
-            let mut a;
-            let mut b;
-            let mut buf = match borrow {
-                Ok(buf) => {
-                    a = buf;
-                    &mut *a
-                }
-                _ => {
-                    b = String::new();
-                    &mut b
-                }
-            };
-            if self.fmt_event.format_event(&ctx, &mut buf, event).is_ok() {
-                let mut writer = self.make_writer.make_writer();
-                if let Err(err) = std::io::Write::write_all(&mut writer, buf.as_bytes()) {
-                    if err.kind() != std::io::ErrorKind::WouldBlock {
-                        let _ = std::io::stderr().write_all(
-                            format!("cannot flush the logging buffer, this is a bug: {err:?}")
-                                .as_bytes(),
-                        );
+            let mut buf = &mut *(buf.borrow_mut());
+            match self.fmt_event.format_event(&ctx, &mut buf, event) {
+                Ok(()) => {
+                    let mut writer = self.make_writer.make_writer();
+                    let limit = buf.len();
+                    let mut index = 0;
+                    let mut carry_on = true;
+                    while index < limit {
+                        // NOTE: Do not raise this limit about 1_024 for macOS without testing it works
+                        // correctly. This figure was determined experimentally by observing pauses in
+                        // logging during startup. If the figure goes above 1_024, you will get a 5s
+                        // pause during logging at startup on macOS. This may change in the future or
+                        // depending on your version of macOS, so we'll need to verify it a couple of
+                        // times a year to see if anything has changed.
+                        #[cfg(target_os = "macos")]
+                        let upper = index + std::cmp::min(limit - index, 1_024);
+                        #[cfg(not(target_os = "macos"))]
+                        let upper = limit;
+
+                        match writer.write(&buf.as_bytes()[index..upper]) {
+                            Ok(written) => {
+                                carry_on = true;
+                                index += written;
+                            }
+                            Err(err) => {
+                                if !carry_on || err.kind() == std::io::ErrorKind::BrokenPipe {
+                                    break;
+                                }
+                                if err.kind() != std::io::ErrorKind::WouldBlock {
+                                    carry_on = false;
+                                    let _ = std::io::stderr().write_all(
+                                        format!("error during write: {err:?}\n").as_bytes(),
+                                    );
+                                    let _ = std::io::stderr().flush();
```

Comment:
maybe add a break here. If we could not write that log we should not insist

Reply:
We try twice (controlled by `carry_on`) and then break whilst ignoring any errors that may occur during writes to `stderr`.
I'm trying to trade off between immediately failing for a temporary out of pipe space type problem and trying too many times. Twice seems reasonable, but it's only an opinion.